### ISSUE TITLE
HybridData: a class to operate and interact with complex data structures

### ISSUE THREAD

**Issue by hgrecco** [napari/napari#4345]
# HybridData: a class to operate and interact with complex data structures

*Draft: 2022-03-26*

## Abstract

This is a proposal to add a HybridData to Napari to hold, operate and display complex data structures that have one (or many) image representation(s).

## Motivation

From the first single intensity 2D image to the current 3D multicolor videos, digital imaging techniques have grown in complexity and data density. Image processing software and applications have improved to display, navigate, annotate, process, analyze and quantify such rich multidimensional stacks. However, not all imaging techniques provide a straight forward or efficient representation as a compact and dense multidimensional array. Some data might be better represented as unstructured linked content, or depend on an internal state that is important for further analysis but might be lost when represented as image. More importantly, the way to explore this type of data usually requires generating certain images on the fly. As convergence between multiple methods such as spectroscopic and imaging techniques becomes more ubiquitous, this proposal hopes to make Napari better suited for this.

## Rationale

We propose adding a `HybridData` class to Napari. A subclass of `HybridData`:

1. must hold a pointer to raw data, file or state, which might be accessible to plugins.
2. must implement NumPy array protocol to provide a default image representation of the data.
3. might provide additional methods (annotated with any Napari types) for other representation of the raw data.
  

Napari will inspect an instance of a `HybridData` subclass and display the default image representation (as returned using NumPy array protocol). Additional image representations might be provided via a context menu or toggle layer.

As a first dummy example we show the idea with a random image that is always generated from the same seed. The random number generator represents here the complex data structure that is more efficiently stored with a seed. This has no utility, just demonstrates the proposal with something that does not require to know the specific details about any technique.

```python
import numpy as np

class RandomImage(HybridData):

    def __init__(self, seed=42, low=0.0, high=1.0, size=(128, 128)):
        self.seed = seed
        self.low = low
        self.high = high
        self.size = size

    def __array__(self, dtype=None):
        return self.as_image()

    def as_image(self) -> ImageData:
        return np.random.default_rng(self.seed).uniform(low, high, size)

    def as_shuffled_image(self) -> ImageData:
        arr = self.as_image()
        np.random.default_rng(seed).shuffle(arr)
        return arr
```

A plug-in might be able to read and process the image representations as well as the raw data; and produce new `HybridData` or other Napari compatible types.

```python
def increase_seed(data: RandomImage, increment: int) -> RandomImage
    return RandomImage(data.seed + increment,
                       data.low,
                       data.high,
                       data.size)
```

## Examples

The following section provides concrete application examples, ranging from some that could be implemented using currently available image layer to others in which this would be more challenging. The actual technique is described very briefly, with an emphasis in the information being used and displayed. We encourage experts in the different methods to add more information if they think is necessary.

### Hyperspectral imaging

Measures the spectra (reflection, transmission, emission) for each position in sample or scene. It has applications in astronomy, remote sensing and biomedical imaging among other. Like any other multicolor image, each wavelength range is a channel. But here, there are between 100 and 1000 different channels. Depending on how the data was acquired (spatial scanning, spectral scanning, spatio-temporal scanning, non scanning) the data might be store and

We chose this as a first example because it is on the edge between a multicolor data image and complex data stream. Everything we mention here could be implemented now in Napari pre-transforming the source data into a (pixels width, pixels height, wavelengths) shaped array plus some plugins. But we think that the approach using `HybridData` could be more convenient.

While navigating each wavelength layer is possible, hyperspectral image hybrid data class might also expose methods to visualize common representations such as an RGB image or those used in remote sensing such as infrared, normalized difference water index, normalized difference vegetation index among others. Plugins might do additional data transformation and quantifications such as finding spectral signatures and making maps of individual components (in a similar way to what is done with spectral unmixing).

### Force spectroscopy in Atomic Force Microscopy (AFM)

In AFM, a tip is brought very near the surface of the sample of interest. Due to the different forces acting on it (e.g. van der Waals, electrostatic) the cantilever holding the tip is deflected. Scanning the tip over the sample provides a topographic/force image of the surface. Additionally, a force vs distance curve for each point can be obtained to study the energy landscape underlying the interaction. Sometimes this is done multiple times not only for statistical purposes but to measure the hysteresis loop. It has applications in material characterization, biomedical sciences and nanotechnology, among other.

A force spectroscopy hybrid data class might expose as a default image the topographic map of the sample, adhesion, dissipation or stiffness. Plugins might do additional data transformations such as determining the presence of hysteresis, etc.

### Energy-dispersive X-ray spectroscopy

Elemental analysis of surfaces in SEM is performed using energy dispersive spectroscopy (EDS), which measures the energy and intensity distribution of X-ray signals generated by the electron beam striking the surface of the specimen. It has application on materials and nanosciences.

An EDS hybrid data class might expose as a default image the SEM image of the samples, as well as additional images for each element.

### Fluorescence Lifetime Imaging Microscopy (FLIM)

Measures the fluorescence lifetime of molecular species to obtain information about their nano-environment. In the time time correlated single photon counting implementation, a pulsed laser is used to periodically excite the sample while its being scanned. Each photon can be tagged with information about the `x`, `y`, `z` position within the sample, as well as the `frame` and `channel`; and most importantly the photon arrival (in the nanosecond range). Molecular species with longer lifetimes will have a photon arrival time histogram with a larger tail. Therefore the data set is not an image but an actual list of photons that can be displayed, filtered and analyzed in multiple ways.

A FLIM hybrid data class might expose the intensity image as the default one, while also exposing others such as the average photon arrival time or photon flux. Plugins might do additional analysis such as IRF inference, lifetime fitting (for which there are multiple ways) or FRET analysis.

## Deferred ideas

**Image representation that depend on additional parameters**

In certain cases, an operations required to calculate an image representations uses values provided by the user.

In the dummy example given above, shuffling n times might look like this.

```python
class RandomImage(HybridData):

    def as_n_shuffled_image(self, n: int) -> ImageData:
        arr = self.as_image()
        for _ in range(n):
            np.random.default_rng(seed).shuffle(arr)
        return arr
```

While we do not see now a problem of supporting this in the future, we think it is currently better to do this type of operations through plugins as the user interactions are easier to implement. Developers can still provide these methods in the class but Napari will not show them as image representations.

## Rejected ideas

**Create an import plugin that calculates the image(s) of choice**

Using an import plugin is a straight forward way to support any type of data file. But is like having a converter script that runs first, dumps images to disk which are then loaded by Napari. This approach lacks interoperability with Napari ecosystem. For example, the analysis it might require to do a background estimation, a morphological operation or region of interest selection; and therefore the import plugin might implement those again. Furthermore, it lacks the ability to go back and forth in the analysis in an easy way.

**Use a sparse matrix and display it a multidimensional array**

It is true that at the very basic, all observables will be acquired in a digital form and therefore a unique indexed value could be store. Time, wavelength, current, voltage, etc are all digitized.. While a sparse multidimensional array is suitable to efficiently store these type of datasets, it is not suitable for all. But most importantly, exploring these datasets might not be in any of the dimensions it has been acquired but in a certain combination of those.

**Non type annotation mechanism for image representation method introspection**

We have considered other ways to let Napari know which methods are image representation methods:

- such as a special naming convention: this usually leads to weird names
- having an extra method/attribute/property that list them all: this involves an additional book-keeping which we think is unnecessary.
- a decorator: this might be ok but we do not see a real benefit.
  

Type annotation are becoming more and more common in Python and IDEs are able to use them to assist during coding and therefore they might be the best choice here.

## Naming

Other that have been considered:

- ObjectData or ObjectLayer: this might be confused with a future 3D shape layer.
- Data: too generic.
  

Bikesheding the name might be necessary, and actually appreciated.

**Comment by hgrecco** [comment]
After an initial conversation with @jni about this, he encouraged me to post it here.

**Comment by jni** [comment]
Thanks @hgrecco for raising this! The code snippet is very handy to think about implementation ideas.

One thing I wonder is whether this can evolve as a separate plugin ecosystem first, then we can have discussions about which bits belong in napari core as usage spreads. Here's one way I can see this playing out:

- First, define HybridData or ComplexData or whatever in some third party package. It should satisfy the [array API](https://data-apis.org/array-api/latest/API_specification/array_object.html) to be viewable in napari by default.
- Then, define a plugin with functions that take in `ImageLayer`s, look at the `.data` attribute of the layer, and, if it's a ComplexData object, expose any annotated methods on the object using magicgui, or just plain "call" buttons if the methods don't take parameters.
- As new uses come up, the community can experiment, find the bits that work and the bits that don't, and we can converge on a generic object of this type
- Which may or may not live in napari proper — to be honest, this seems "bigger" than napari, and we just want to make sure that napari plays well with it!

What do you think? What are blockers to doing that that I haven't caught?

**Comment by hgrecco** [comment]
@jni this is a reasonable proposal. I still think that it would be nice to provide a more native way to display alternative visualizations, but this can come at a later stage. That is what I would include in Napari, while allowing the rest to leave outside.

**Comment by jni** [comment]
Yup yup! I can see that, but as we have a long list of things to implement (even axis naming and axis ordering can't be done in the UI yet!), and this one has many design questions open that could stall discussion, I wanted to open a path to getting all the work done outside of napari (where it can move faster), and then move it in. Obviously, if there are aspects that *must* happen in napari, and they become apparent during the work, then we can address those as they happen!
